rmw_connext_cpp/src/functions.cpp

```diff
@@ -1630,4 +1846,246 @@ rmw_send_response(
+    }
+    if (topic_names_and_types->topic_names) {
+      rmw_free(topic_names_and_types->topic_names);
+      topic_names_and_types->topic_names = nullptr;
+    }
+    if (topic_names_and_types->type_names) {
+      rmw_free(topic_names_and_types->type_names);
+      topic_names_and_types->type_names = nullptr;
+    }
+    topic_names_and_types->topic_count = 0;
+  }
+}
+
+rmw_ret_t
+rmw_get_topic_names_and_types(
+  const rmw_node_t * node,
+  rmw_topic_names_and_types_t * topic_names_and_types)
+{
+  if (!node) {
+    RMW_SET_ERROR_MSG("node handle is null");
+    return RMW_RET_ERROR;
+  }
+  if (node->implementation_identifier != rti_connext_identifier) {
+    RMW_SET_ERROR_MSG("node handle is not from this rmw implementation");
+    return RMW_RET_ERROR;
+  }
+  if (!topic_names_and_types) {
+    RMW_SET_ERROR_MSG("topics handle is null");
+    return RMW_RET_ERROR;
+  }
+  if (topic_names_and_types->topic_count) {
+    RMW_SET_ERROR_MSG("topic count is not zero");
+    return RMW_RET_ERROR;
+  }
+  if (topic_names_and_types->topic_names) {
+    RMW_SET_ERROR_MSG("topic names is not null");
+    return RMW_RET_ERROR;
+  }
+  if (topic_names_and_types->type_names) {
+    RMW_SET_ERROR_MSG("type names is not null");
+    return RMW_RET_ERROR;
+  }
+
+  auto node_info = static_cast<ConnextStaticNodeInfo *>(node->data);
+  if (!node_info) {
+    RMW_SET_ERROR_MSG("node info handle is null");
+    return RMW_RET_ERROR;
+  }
+  if (!node_info->publisher_listener) {
+    RMW_SET_ERROR_MSG("publisher listener handle is null");
+    return RMW_RET_ERROR;
+  }
+  if (!node_info->subscriber_listener) {
+    RMW_SET_ERROR_MSG("subscriber listener handle is null");
+    return RMW_RET_ERROR;
+  }
+
+  // combine publisher and subscriber information
+  std::map<std::string, std::set<std::string>> topics_with_multiple_types;
+  for (auto it : node_info->publisher_listener->topic_names_and_types) {
+    for (auto jt : it.second) {
+      topics_with_multiple_types[it.first].insert(jt);
+    }
+  }
+  for (auto it : node_info->subscriber_listener->topic_names_and_types) {
+    for (auto jt : it.second) {
+      topics_with_multiple_types[it.first].insert(jt);
+    }
+  }
+
+  // ignore inconsistent types
+  std::map<std::string, std::string> topics;
+  for (auto it : topics_with_multiple_types) {
+    if (it.second.size() != 1) {
+      fprintf(stderr, "topic type mismatch - ignoring topic '%s'\n", it.first.c_str());
+      continue;
+    }
+    topics[it.first] = *it.second.begin();
+  }
+
+  // reformat type name
+  std::string substr = "::msg::dds_::";
+  for (auto it : topics) {
+    size_t substr_pos = it.second.find(substr);
+    if (it.second[it.second.size() - 1] == '_' && substr_pos != std::string::npos)
+    {
+      it.second = it.second.substr(0, substr_pos) + "/" +
+        it.second.substr(substr_pos + substr.size(), it.second.size() - substr_pos - substr.size() - 1);
+      topics[it.first] = it.second;
```

Comment:
I don't think this should be necessary if you get a reference to the pair from the `topics` map.